src/middlewared/middlewared/plugins/auth.py

```diff
@@ -325,9 +339,130 @@ def token(self, app, token):
+        `otp_digits` represents number of allowed digits in the OTP.
+
+        `window` extends the validity to `window` many counter ticks before and after the current one.
+
+        `interval` is time duration in seconds specifying OTP expiration time from it's creation time.
+        """
+        old_config = await self.config()
+        config = old_config.copy()
+
+        config.update(data)
+
+        if config['enabled'] and not config['secret']:
+            # Only generate a new secret on `enabled` when `secret` is not already set.
+            # This will aid users not setting secret up again on their mobiles.
+            config['secret'] = await self.middleware.run_in_thread(
+                self.generate_base32_secret
+            )
+
+        config['secret'] = await self.middleware.call('pwenc.encrypt', config['secret'])
+        config.pop('secret_hex', None)
+
+        await self.middleware.call(
+            'datastore.update',
+            self._config.datastore,
+            config['id'],
+            config
+        )
+
+        await self.middleware.call('service.reload', 'ssh')
+
+        return await self.config()
+
+    @accepts(
+        Str('token', null=True)
+    )
+    def verify(self, token):
+        """
+        Returns boolean true if provided `token` is successfully authenticated.
+        """
+        config = self.middleware.call_sync(f'{self._config.namespace}.config')
+        if not config['enabled']:
+            raise CallError('Please enable Two Factor Authentication first.')
+
+        totp = pyotp.totp.TOTP(
+            config['secret'], interval=config['interval'], digits=config['otp_digits']
+        )
+        return totp.verify(token, valid_window=config['window'])
+
+    @accepts()
+    def renew_secret(self):
+        """
+        Generates a new secret for Two Factor Authentication. Returns boolean true on success.
+        """
+        config = self.middleware.call_sync(f'{self._config.namespace}.config')
+        if not config['enabled']:
+            raise CallError('Please enable Two Factor Authentication first.')
+
+        self.middleware.call_sync(
+            'datastore.update',
+            self._config.datastore,
+            config['id'], {
+                'secret': self.middleware.call_sync('pwenc.encrypt', self.generate_base32_secret())
+            }
+        )
+
+        if config['services']['ssh']:
+            self.middleware.call_sync('service.reload', 'ssh')
+
+        return True
+
+    @accepts
+    def provisioning_uri(self):
+        """
+        Returns the provisioning URI for the OTP. This can then be encoded in a QR Code and used to
+        provision an OTP app like Google Authenticator.
+        """
+        config = self.middleware.call_sync(f'{self._config.namespace}.config')
+        return pyotp.totp.TOTP(
+            config['secret'], interval=config['interval'], digits=config['otp_digits']
+        ).provisioning_uri('root@freenas', 'iXsystems')
```

Comment:
How does this work and why are we hardcoding these values?

Reply:
So first param is `name` which is of the user account which is being secured, second is the name of the provider/issuer name which will show up in the authenticator app as the organization which i defaulted to `iXsystems`. These values don't have any role to play in the authentication itself but are used by the 2FA apps to help users know which code is being used for which service etc.
We can perhaps allow this to be configurable - the account part ?
https://github.com/google/google-authenticator/wiki/Key-Uri-Format

Reply:
I think we should at least use the hostname of freenas to make it more identifiable then